tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala

```diff
@@ -54,14 +57,53 @@ object CodeGen {
       // of available processesors. This prevents LLVM from optimizing
       // across IR module boundary unless LTO is turned on.
       def separate(): Seq[Path] =
-        partitionBy(assembly, procs)(_.name.top.mangle).par
+        assembly
+          .groupBy { defn =>
+            val packageName =
+              defn.name.top.id.split("\\.").dropRight(1).mkString(".")
+            packageName
+          }
+          .par
           .map {
-            case (id, defns) =>
-              val sorted = defns.sortBy(_.name.show)
-              Impl(config, env, sorted).gen(id.toString, workdir)
+            case (pack, defns) =>
+              if (incCompilationContext != null) {
+                incCompilationContext.collectFromCurr(pack, defns)
+              }
+              if (incCompilationContext == null ||
+                  incCompilationContext.isChanged(pack)) {
+                val sorted = defns.sortBy(_.name.show)
+                val packagePrefix = config.workdir resolve
+                  Paths.get(
```

Comment:
resolve also accepts String as a parameter, so we can skip pass it directory. In case of multiline method calls please use dots and braces